[script] [forging] Adding support for adjustable tongs

Raykyn55 opened the pull request:
Checks for adjustable tongs, if they adjust, uses them as shovel, then reverts before next crafting action. Also corrects tongs in wrong adjustment during pound.

vtcifer commented:
II would recommend a new yaml setting to enable this, rather than assuming everyone uses adjustable tongs, otherwise there's a bit of a regression of functionality here.

Raykyn55 commented:
Still works for regular tongs

SekmehtDR commented:
Nice job putting something together! I see where VTC is coming from. Could have it do a get_settings for swappable_tongs: true, then if true, then do the adjustment, if not then not even check it. 

I'd like to say most of everyone who has high end tongs, or covellite articulated tongs even, all of those are basically swappable, so the assumption part is semi-valid.

Raykyn55 commented:
OK, so I can make this a yaml setting, but didn't for a couple reasons (though absolutely willing to go that route)
1. The changes I made here trigger a quick attempt at adjustment maybe once per item, and if it fails, it just carries on getting a shovel. So no one loses function.
2. The yaml is already pretty big, and I didn't want to add more niche settings where something like this could suffice (IMO)
3. When testing, i discovered that use of the shovel is pretty rare(to my dismay, prob went through 20 hand claws to get 5 tests)

That said, I am working on a yaml setting version, how would I submit that? Put it in here?

SekmehtDR commented:
I like the quick adjust and match with it, if it adjusts then it can be used as shovel then returned... Makes since to me right?

MahtraDR commented:
> That said, I am working on a yaml setting version, how would I submit that? Put it in here?

You'd add it to this PR.

SekmehtDR commented:
> 

Yeah and like Mahtra said, when you make any changes to say base.yaml, base will be listed as another file changed as long as it's all within the same branch :)

SekmehtDR commented:
@MahtraDR in the title of this PR, maybe remove clockwork, because there's actually a lot of adjustable tongs in the game, the craft able one being articulated tongs. So I'd make it "... for adjustable tongs"

MahtraDR commented:
> @MahtraDR in the title of this PR, maybe remove clockwork, because there's actually a lot of adjustable tongs in the game, the craft able one being articulated tongs. So I'd make it "... for adjustable tongs". But hell yeah im using my clockworks with it :D

I did not know there was more than one kind of adjustable tongs!

SekmehtDR commented:
> I did not know there was more than one kind of adjustable tongs!

Cool right!?! swappable between shovel and tong and retains the speed!

MahtraDR commented on forge.lic:
```diff
@@ -284,6 +284,7 @@ class Forge
       pound(item)
     when /You must be holding some metal tongs in your other hand to do that/
       get_item('tongs')
+      bput("adjust my tongs", 'alongside the tong', 'You cannot adjust')
```
```suggestion
      DRC.bput("adjust my tongs", 'alongside the tong', 'You cannot adjust')
```

MahtraDR commented on forge.lic:
```diff
@@ -321,16 +322,28 @@ class Forge
-      stow_item(@shovel)
-      get_item('tongs')
-      pound
+    case bput("adjust my tongs", 'alongside the tong', 'You cannot adjust', 'You have no idea how')
```
```suggestion
    case DRC.bput("adjust my tongs", 'alongside the tong', 'You cannot adjust', 'You have no idea how')
```

MahtraDR commented on forge.lic:
```diff
@@ -321,16 +322,28 @@ class Forge
+    when 'You cannot adjust', 'You have no idea how'
+      stow_item('tongs')
+      get_item(@shovel)
+      case bput('push fuel with my shovel', 'Roundtime', 'That tool does not seem')
```
```suggestion
      case DRC.bput('push fuel with my shovel', 'Roundtime', 'That tool does not seem')
```

MahtraDR commented on forge.lic:
```diff
@@ -321,16 +322,28 @@ class Forge
+        pound
+      end
+    when /alongside the tong/
+      case bput('push fuel with my tongs', 'Roundtime', 'That tool does not seem')
```
```suggestion
      case DRC.bput('push fuel with my tongs', 'Roundtime', 'That tool does not seem')
```

MahtraDR commented on forge.lic:
```diff
@@ -321,16 +322,28 @@ class Forge
+        analyze_item
+      else
+        waitrt?
+        bput("adjust my tongs", 'alongside the tong')
```
```suggestion
        DRC.bput("adjust my tongs", 'alongside the tong')
```

MahtraDR reviewed (changes requested):
Please use explicit prefixes for the common methods. I've suggested the change on all the bputs.

Raykyn55 commented:
Ok, added those DRCs, removed includes, and made it a yaml setting: adjustable_tongs

Raykyn55 commented:
false in base, blank in empty

SekmehtDR commented:
`R> 
[forge]>pound fan on anvil with my forging hammer

After using the tongs to warm the fan over the forge fire, you place it down upon the anvil and make some adjustments to its shape with carefully timed hammer taps.  The forging proceeds exceptionally well and free of mistakes.

As you complete working the fire dies down and appears to need some more fuel.
Roundtime: 9 sec.

R> 
Gained: Forging(+1)

R> 
[forge]>adjust my tongs

You lock the tongs into a fully extended position.  Then you reach alongside the tong's body, and pull free the head of a shovel.

> 
[forge]>push fuel with my tongs

Gripping the tongs tightly, you slam it deep into the pile of fuel and give it a hard kick.   Straining, you remove the tongs with a mound of coal atop it and toss the coal onto the fire.
Roundtime: 10 sec.

R> 
[forge]>adjust my tongs

With a yank you fold the shovel head back alongside the tong's body.  Then you unlock the arms, readying them for use as tongs once more.

> 
[forge]>pound fan on anvil with my forging hammer

Following solid heating in the forge, you carefully fold the glowing red metal back onto itself and hammer it until both ends are fused.  Blow after blow rings sound without any problems at all.
Roundtime: 8 sec.
`

Tested and confirms the adjusting is working flawlessly with the **adjustable_tongs: true** yaml setting in place, then when set to false, the prior behavior swapping to shovel was noted.

I have not tested smelt.lic yet.

Nice job! :)

MahtraDR commented:
Closes https://github.com/rpherbig/dr-scripts/issues/2161

MahtraDR commented:
I appreciate the effort at removing the includes, but we have a lot of prefixes missing if we do that. 

![image](https://user-images.githubusercontent.com/93822896/165445376-6fdb8c79-c175-4bf9-9ba9-cc44d390a3b6.png)

In red is `DRC`, in orange is `DRCC` in pink is `DRCA` and so on.  I have a VSCode addon I use to highlight like that, if you want to use it.

Raykyn55 commented:
Changed everything but the get_item/stow_item, as it references a method in the script. Should thoose methods be removed, and the get_item be changed to DRCC.get_crafting_item/stow?

Raykyn55 commented:
Ok, did a bunch of work this morning on both forge and smelt, cut some fat, sped up grinding, added your prefixes, split fuel into two methods. I'm posting here, but wondering if i should start a fresh one?  Let me know.

Raykyn55 commented on forge.lic:
```diff
@@ -2,7 +2,7 @@
 =end
 
-custom_require.call(%w[common common-arcana common-crafting common-travel events drinfomon])
+custom_require.call(%w[common common-arcana common-crafting common-money common-travel events drinfomon])
```
common-money was added mistakenly, forgot to remove.

MahtraDR commented:
This depends on https://github.com/rpherbig/dr-scripts/pull/5767

asechrest commented:
> OK, so I can make this a yaml setting, but didn't for a couple reasons (though absolutely willing to go that route)

> 2. The yaml is already pretty big, and I didn't want to add more niche settings where something like this could suffice (IMO)

Note that I'm working on a substantial set of changes that will help users run help output on any script and see the yaml settings used in the script. You're right that `base.yaml` is becoming unwieldy, so that should help.

MahtraDR reviewed (approved):
This version I am quite happy with. 

It depends on some changes to common-crafting which were merged today. 

We'll merge this in 48ish hours to give those changes a chance to propagate.

Dartellum commented:
#5065 has this already.

MahtraDR reviewed (approved):
Tested over the last few days. Looks good.
